Update Oracle library to latest version

lmeerwood opened the pull request:
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

Updated the Oracle library from the old [cx_oracle](https://oracle.github.io/python-cx_Oracle/) library to the new [oracledb](https://oracle.github.io/python-oracledb/) library. Technically they're the same library, it was just renamed recently.

One massive difference between the two versions is that oracledb no longer requires the user to have Oracle Client libraries installed on their machine, the library can connect directly to the server in thin client mode. This should make it easier to setup a connection to Oracle Databases. 

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragement file, named `{pr_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

Commit:
Replaced cx_oracle with new oracledb library

lmeerwood committed:
Merge branch 'apache:main' into main

Commit:
Removed errant tab whitespace

Commit:
Changes recommended by static-test

Commit:
Fixed line length

Commit:
Fixed type import

Commit:
Fixed mypy issue

Commit:
Applied hack suggested by
https://github.com/python/mypy/issues/1297#issuecomment-508593494

Commit:
Merge branch 'main' of github.com:apache/airflow into main

Commit:
Removed errant print statement

dungdm93 commented on airflow/providers/oracle/hooks/oracle.py:
```diff
@@ -135,35 +135,35 @@ def get_conn(self) -> 'OracleHook':
+            conn_config['purity'] = oracledb.ATTR_PURITY_SELF
         elif purity == 'default':
-            conn_config['purity'] = cx_Oracle.ATTR_PURITY_DEFAULT
+            conn_config['purity'] = oracledb.ATTR_PURITY_DEFAULT
```
`ATTR_PURITY_*` are aliases for backwards compatibility.
I think it's better to use the new one:
```python
oracledb.PURITY_DEFAULT
oracledb.PURITY_NEW
oracledb.PURITY_SELF
```

dungdm93 commented on airflow/providers/oracle/hooks/oracle.py:
```diff
@@ -135,35 +135,35 @@ def get_conn(self) -> 'OracleHook':
+            conn_config['mode'] = oracledb.SYSKMT
         elif mode == 'sysrac':
-            conn_config['mode'] = cx_Oracle.SYSRAC
+            conn_config['mode'] = oracledb.SYSRAC
```
Same as `ATTR_PURITY_*`, all thoses props are aliases for backwards compatibility.
I think it's better to use new props `oracledb.AUTH_MODE_*`

dungdm93 commented on airflow/providers/oracle/hooks/oracle.py:
```diff
@@ -123,7 +123,7 @@ def get_conn(self) -> 'OracleHook':
             # if `encoding` is specific but `nencoding` is not
             # `nencoding` should use same values as `encoding` to set encoding, inspired by
-            # https://github.com/oracle/python-cx_Oracle/issues/157#issuecomment-371877993
+            # https://github.com/oracle/python-oracledb/issues/157#issuecomment-371877993
```
`encoding`, `nencoding` and `threaded` are deprecated and has no effect.
It's safe to remove.

See: 
* https://github.com/oracle/python-oracledb/blob/2068d83baba2c18164285e63b9a16e16e5f9b219/src/oracledb/connection.py#L1047-L1050
* https://github.com/oracle/python-oracledb/blob/2068d83baba2c18164285e63b9a16e16e5f9b219/src/oracledb/connection.py#L393

dungdm93 commented on airflow/providers/google/cloud/transfers/oracle_to_gcs.py:
```diff
@@ -45,15 +45,15 @@ class OracleToGCSOperator(BaseSQLToGCSOperator):
+        oracledbbase.DB_TYPE_NUMBER: 'NUMERIC',
+        oracledbbase.DB_TYPE_TIMESTAMP: 'TIMESTAMP',
+        oracledbbase.DB_TYPE_TIMESTAMP_LTZ: 'TIMESTAMP',
+        oracledbbase.DB_TYPE_TIMESTAMP_TZ: 'TIMESTAMP',
```
I do not recommend direct using underlay implementation of any package. Is there any reason why not using:
```python
import oracledb

oracledb.DB_TYPE_TIMESTAMP_TZ: 'TIMESTAMP'
```

lmeerwood replied:
I had that originally, but MyPy didn't like that. I'd get the following error:
`error: Module has no attribute DB_TYPE_TIMESTAMP_TZ: 'TIMESTAMP'`. This was the only way I could find to stop the error.

lmeerwood replied:
@potiuk, I feel like my solution here is a little bit hacky. The only problem I have with doing it the way @dungdm93 suggest is that it causes issues with MyPy, stating the error I wrote above. I did try the fix you mentioned in Slack (`breeze build-image --upgrade-to-newer-depenencies --python 3.7`) but that did not solve the issue.

You also suggested for another issue I had with MyPy that some times ignoring a specific typing error is the easiest approach. Do you think this might also be the case for the issue?

Commit:
Removed usage of backwards compatible names.
Removed deprecated encoding, nencoding and threaded.
Removed hack for import and used mypy ignore comment instead
Removed use of underlying implementation and used mypy ignore comment instead

dungdm93 commented on tests/providers/oracle/hooks/test_oracle.py:
```diff
@@ -158,12 +158,12 @@ def test_get_conn_events(self, mock_connect):
-            'default': cx_Oracle.ATTR_PURITY_DEFAULT,
+            'new': oracledb.ATTR_PURITY_NEW,
+            'self': oracledb.ATTR_PURITY_SELF,
+            'default': oracledb.ATTR_PURITY_DEFAULT,
```
Should be `oracledb.PURITY_*` as well

dungdm93 commented on tests/providers/google/cloud/transfers/test_oracle_to_gcs.py:
```diff
@@ -19,7 +19,7 @@
 from unittest import mock
 
-import cx_Oracle
+import oracledb.base_impl as oracledbbase
```
Should import `oracledb` as well

dungdm93 commented on tests/providers/oracle/hooks/test_oracle.py:
```diff
@@ -119,15 +119,15 @@ def test_get_conn_nencoding(self, mock_connect):
+            'sysoper': oracledb.SYSOPER,
+            'sysbkp': oracledb.SYSBKP,
+            'sysdgd': oracledb.SYSDGD,
+            'syskmt': oracledb.SYSKMT,
```
Should using `oracledb.AUTH_MODE_*` as well

Commit:
Removed use of aliases.

Commit:
Removed tests relating to deprecated features.
Fixed typo.
Commited changes to conf.py mandated by breeze

potiuk reviewed (approved):
Cool. looks great!
